### PR TITLE
lkl: Complain more in virtio_process_one

### PR DESCRIPTION
**Pull request by pscollins**
Following the example of tools/lguest/lguest.c, a vring descriptor
chain that is too long to fit into our request is an error condition, so
we warn about it and panic when it happens.

This commit also adds some documentation to the very dense LKL-side virtio
code and splits up virtio_process_one into smaller functions.

Signed-off-by: Patrick Collins pscollins@google.com

[Review on Reviewable](https://reviewable.io/reviews/lkl/linux/74)

### CODE REVIEW

File: tools/lkl/lib/virtio.c [hunk 1]
```diff
@@ -82,30 +82,54 @@ void virtio_req_complete(struct virtio_req *req, uint32_t len)
 		virtio_deliver_irq(dev);
 }
 
+/* Grab the vring_desc from the queue at the appropriate index in the
+ * queue's circular buffer, converting from little-endian to
+ * the host's endianness. */
+struct lkl_vring_desc *vring_desc_at_le_idx(struct virtio_queue *q, __lkl__virtio16 le_idx)
```
Comment (stfairy) [inline tools/lkl/lib/virtio.c]: static inline

**Comment by pscollins** [comment]
@stfairy Done

**Comment by opurdila** [comment]
Hi Patrick,

I have intentionally did not crash in case we have more buffers so that we can still have a chance to handle the request in the upper layers.

I was planning to add new APIs to allow the upper layers get the buffers that do not fit in the request structure but did not get to it yet.

**Comment by pscollins** [comment]
@opurdila Okay, makes sense. I wrapped the warning/panic it inside of an `#ifdef DEBUG`, which seems to be common among other virtio-related code I've seen. Is that alright, or do you still think it's messy/unnecessary?

**Comment by pscollins** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[tools/lkl/lib/virtio.c, line 88 \[r1\]](https://reviewable.io:443/reviews/lkl/linux/74#-KAgSkyQxgEA551QxaQP-r1-88:-KAkheQKeAQyXof22Ymy:-499146009) ([raw file](https://github.com/lkl/linux/blob/ab9687630f0addd7ee36d422b3d68342bce045b7/tools/lkl/lib/virtio.c#L88)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/lkl/linux/74)*

**Comment by opurdila** [comment]
I think we should do the checking in the upper drivers (virtio_net and virtio_blk) since at that point we know how many buffers we expect. 

I will merge this as it is for now since there are lots of PR I was not able to merge in a week now :) but if you agree with my statement above lets log an issue to track that.

**Comment by pscollins** [comment]
Sure, I'll open it tomorrow. I don't have any strong opinion about where
problems should actually be dealt with, I just want future people debugging
virtio code to be able to easily tell that:

* this is one way that this code can go wrong, and

* they don't need to worry about it, because the print statement hasn't
fired

**Pull request merged**